### PR TITLE
Fix Window operator with NaN as the frame bound

### PR DESCRIPTION
**Pull request by kagamiori**
Summary:
NaN could appear as the frame bound of k-range frames in window 
operations. When NaN appear in either of the frame bounds, the window 
operation should return NULL at this row.

This diff fixes https://github.com/facebookincubator/velox/issues/11213.

Differential Revision: D64510519

### CODE REVIEW

**Comment by aditi-pandit** [comment]
@kagamiori : Quick question : This logic should be applicable for k row frames as well. If yes, please modify https://github.com/facebookincubator/velox/blob/main/velox/exec/Window.cpp#L347 as well.

**Comment by kagamiori** [comment]
> @kagamiori : Quick question : This logic should be applicable for k row frames as well. If yes, please modify https://github.com/facebookincubator/velox/blob/main/velox/exec/Window.cpp#L347 as well.

Hi @aditi-pandit, NaN shouldn't be an issue to k rows frames because rows frames require the frame bounds to be integral while NaN is floating-point. Let me know if I misunderstand your question.

**Comment by aditi-pandit** [comment]
> > @kagamiori : Quick question : This logic should be applicable for k row frames as well. If yes, please modify https://github.com/facebookincubator/velox/blob/main/velox/exec/Window.cpp#L347 as well.
> 
> Hi @aditi-pandit, NaN shouldn't be an issue to k rows frames because rows frames require the frame bounds to be integral while NaN is floating-point. Let me know if I misunderstand your question.

@kagamiori : Makes sense about k row frame.

Though, we should add a validation in WindowNode for rejecting NaN value for k range constant frame as well https://github.com/facebookincubator/velox/blob/main/velox/core/PlanNode.cpp#L1412

File: velox/exec/WindowPartition.cpp
Comment (aditi-pandit) [inline velox/exec/WindowPartition.cpp:389]: We could avoid this check on a per row basis. They can be computed outside the loop or the outer function could pass to the inner function if it should do a float or double NaN check.

File: velox/exec/RowContainer.h
Comment (xiaoxmeng) [inline velox/exec/RowContainer.h:781]: nit: const auto

File: velox/exec/WindowPartition.h
Comment (xiaoxmeng) [inline velox/exec/WindowPartition.h:161]: Is the comment correct? validFrames?

File: velox/exec/WindowPartition.cpp
Comment (xiaoxmeng) [inline velox/exec/WindowPartition.cpp:387]: Can we directly call this method and return false for non floating data type?

**Review by xiaoxmeng — COMMENTED** [review]
@kagamiori thanks for the fix % minors!

**Comment by kagamiori** [comment]
> > > > @kagamiori : Quick question : This logic should be applicable for k row frames as well. If yes, please modify https://github.com/facebookincubator/velox/blob/main/velox/exec/Window.cpp#L347 as well.
> > > 
> > > 
> > > Hi @aditi-pandit, NaN shouldn't be an issue to k rows frames because rows frames require the frame bounds to be integral while NaN is floating-point. Let me know if I misunderstand your question.
> > 
> > 
> > @kagamiori : Makes sense about k row frame.
> > Though, we should add a validation in WindowNode for rejecting NaN value for k range constant frame as well https://github.com/facebookincubator/velox/blob/main/velox/core/PlanNode.cpp#L1412
> 
> @kagamiori : I realized later that k range frames (even if constant) always have range frame column. So we don't get the constant k value in WindowNode PlanNode frame. So it won't be possible for us to reject this case in Velox. We should error it in the co-ordinator. The best at runtime is to return NULL values for the entire result.
> 
> Ideally we shouldn't have such an inconsistent implementation. But we will need to do more design work if we want to do this validation.

Hi @aditi-pandit, just to make sure we're on the same page, according to the discussion in https://github.com/prestodb/presto/issues/23822, we're going to throw an error in Velox if the k range frame column contains a NaN, right?

**Comment by aditi-pandit** [comment]
@kagamiori : Yes. Let me elaborate to reduce confusion. We have 2 error situations with range frames and Nan :

- Constant bound k range frame : (window_func) OVER PARTITION BY ... ORDER BY ... RANGE BETWEEN  NaN PRECEDING AND 5 FOLLOWING
   For such frames Presto co-ordinator computes the range frame column, which contains the value that we have to search for in ORDER BY column for the frame bound. This range frame column is computed as (order_by_column_value - NaN) at each row for the above frame. 
   Generally we want to reject this entire frame bound itself as the range frame column value for each row is not meaningful.
   Velox never sees the original frame string. Velox API only has a spec for the range frame column, which Presto has populated with NaN at each row.
   
   So we will have to rely on Presto rejecting this frame.

- Column bound k range frames : (window_func) OVER PARTITION BY ... ORDER BY ... RANGE BETWEEN  column PRECEDING AND 5 FOLLOWING
  The range frame column is computed for this frame also. The range frame value at any row is order by column value (-/+) range frame column value depending on the semantics of the frame.
  In this situation we can expect that some values could be NaN. This PR is modifying the runtime to always return NULL for such rows. Which is reasonable behavior.
  
Its possible that all column values in the range frame column are NaN and so all window values returned at NULL as well, but that is expected behavior.

To me that is a point of inconsistency between the RANGE frame behavior between the constant and column bound as something that looks similar has different results. We should try to avoid it in general.

But the only option to fix it would be to pass the original frame value to Velox. Definitely something we can do with some design work.

**Comment by kagamiori** [comment]
Hi @aditi-pandit, thank you for the reply! Could you help me understand this?
> Its possible that all column values in the range frame column are NaN and so all window values returned at NULL as well, but that is expected behavior.

I thought that for `(window_func) OVER PARTITION BY ... ORDER BY ... RANGE BETWEEN column PRECEDING AND 5 FOLLOWING`, if there are NaNs in `column`, Presto should reject it and throw an error too before sending the query plan to Velox. (My understanding is that both constant-bound and column-bound k range frame are processed by Presto co-ordinator first and NaN frame bounds can be rejected there, so Velox simply would not see NaNs in k range frame bounds when used in Prestissimo.) Am I missing anything?

**Comment by aditi-pandit** [comment]
> Hi @aditi-pandit, thank you for the reply! Could you help me understand this?
> 
> > Its possible that all column values in the range frame column are NaN and so all window values returned at NULL as well, but that is expected behavior.
> 
> I thought that for `(window_func) OVER PARTITION BY ... ORDER BY ... RANGE BETWEEN column PRECEDING AND 5 FOLLOWING`, if there are NaNs in `column`, Presto should reject it and throw an error too before sending the query plan to Velox. (My understanding is that both constant-bound and column-bound k range frame are processed by Presto co-ordinator first and NaN frame bounds can be rejected there, so Velox simply would not see NaNs in k range frame bounds when used in Prestissimo.) Am I missing anything?
> 
> ==> Update: I've just realized...is it that Presto co-ordinator wouldn't process column bound of k range frames, but would add project node before the window node in the query plan to do that process in workers?

@kagamiori :: Yes. Presto co-ordinator adds project node. It doesn't look at the column values.

**Comment by kagamiori** [comment]
Hi @aditi-pandit and @xiaoxmeng, thank you for reviewing! I've updated the PR. Please take another look. Thanks!

File: velox/exec/WindowPartition.cpp
Comment (aditi-pandit) [inline velox/exec/WindowPartition.cpp:375]: @kagamiori : We don't need these loops anymore.

File: velox/exec/RowContainer.h
Comment (aditi-pandit) [inline velox/exec/RowContainer.h:780]: Do you want to pass RowColumn here instead of columnIndex ? Would avoid one indirection per row.

File: velox/exec/RowContainer.h
Comment (aditi-pandit) [inline velox/exec/RowContainer.h:782]: In the most general sense its better to do the Null check here. 

Though in our case, null check happens earlier anyways. So we are guaranteed non-null values. Do you want to have an optimized version of this code that doesn't have the null check ?

File: velox/exec/tests/WindowTest.cpp
Comment (aditi-pandit) [inline velox/exec/tests/WindowTest.cpp:689]: The offset columns are always corresponding to the same row or NaN. Would be great to make the values more interesting and include some non-Nan values as well.

File: velox/exec/WindowPartition.cpp
Comment (kagamiori) [reply velox/exec/WindowPartition.cpp:375]: Oh, right. Let me remove them.

File: velox/exec/RowContainer.h
Comment (aditi-pandit) [inline velox/exec/RowContainer.h:781]: Nit : Use const& RowColumn parameter.

File: velox/exec/RowContainer.h
Comment (aditi-pandit) [inline velox/exec/RowContainer.h:789]: Nit : Use const& RowColumn parameter.

File: velox/exec/WindowPartition.cpp
Comment (aditi-pandit) [inline velox/exec/WindowPartition.cpp:389]: My bad that we do null check before NaN check. I'm surprised that didn't get caught in the testing.

The NULL check happens in the next if condition that compares for the equality of RowContainer values.  Since NaN != NaN, we could move the unguarded NaN check in the else side of the condition below.

Comment (kagamiori) [reply velox/exec/WindowPartition.cpp:389]: Hi @aditi-pandit, I got what you mean. But it looks like we cannot avoid the check for NULL in RowContainer::isNanAt even in the else branch here? The reason is that RowContainer::compare returns non-zero if frame bound is NULL but order-by value is not NULL. Also, RowContainer::compare actually considers NaN == NaN.

Comment (aditi-pandit) [reply velox/exec/WindowPartition.cpp:389]: @kagamiori : I see. Guess we have to use the version of the NaN check with the NULL check. We can do the check before the DCHECK of the NULL column values at line 378. Thanks.

Comment (kagamiori) [reply velox/exec/WindowPartition.cpp]: Hi @aditi-pandit, I've updated the PR. Please take another look. Thank you!

File: velox/exec/tests/WindowTest.cpp
Comment (aditi-pandit) [inline velox/exec/tests/WindowTest.cpp:733]: @kagamiori : rank() is not affected by frame values. Only value functions (first_value, last_value, nth_value) and aggregate functions are affected by frame boundaries. MIght be better to use one of those functions in the test.

File: velox/exec/WindowPartition.cpp
Comment (aditi-pandit) [reply velox/exec/WindowPartition.cpp]: @kagamiori : Your code looks fine to me except for the test comment below. I'll approve after the test fix.

File: velox/exec/tests/WindowTest.cpp
Comment (kagamiori) [reply velox/exec/tests/WindowTest.cpp:733]: Hi @aditi-pandit, I intentionally added the test case of rank() to ensure it is not affected by the handling of NaN frame bound as you point out. The regular case is tested by the sum() function above.

Comment (aditi-pandit) [reply velox/exec/tests/WindowTest.cpp:733]: @kagamiori : I see. Maybe then add a comment to indicate this is for a function not affected by the Nan frames.

**Review by aditi-pandit — APPROVED** [review]
Thanks @kagamiori for the iterations.

File: velox/exec/tests/WindowTest.cpp
Comment (kagamiori) [reply velox/exec/tests/WindowTest.cpp:733]: @aditi-pandit, sure, will do. Thanks for the suggestion!

File: velox/exec/WindowPartition.cpp
Comment (xiaoxmeng) [inline velox/exec/WindowPartition.cpp:468]: Can we use VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH_ALL here instead of passing void?

Comment (xiaoxmeng) [inline velox/exec/WindowPartition.cpp:468]: Can we use VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH_ALL here instead of passing void?

Comment (xiaoxmeng) [inline velox/exec/WindowPartition.cpp:468]: Can we use VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH_ALL here instead of passing void for non-float or double type?

**Review by xiaoxmeng — COMMENTED** [review]
Thanks for the update!

File: velox/exec/WindowPartition.cpp
Comment (kagamiori) [reply velox/exec/WindowPartition.cpp:468]: Hi @xiaoxmeng, I intentionally attempted to avoid using VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH_ALL. This will cause the code instantiation for every type, but we don't really differentiate the code for non-floating-point types.

File: velox/exec/WindowPartition.cpp
Comment (xiaoxmeng) [inline velox/exec/WindowPartition.cpp:380]: Add a comment to describe this edge case as chat offline? thanks!

**Review by xiaoxmeng — APPROVED** [review]
@kagamiori thanks for the offline explanation!

**Review by xiaoxmeng — COMMENTED** [review]
Consider to add a unit test to cover the new row container API as well as have a test to cover both frame bound and order by columns are nan case? thanks!

**Comment by kagamiori** [comment]
Hi @aditi-pandit, I found that if the order-by value is NaN, it would cause the frame bounds to be NaN too, even if the original frame offsets are not NaN. E.g.,
```
SELECT SUM(c0) OVER (ORDER BY s0 RANGE BETWEEN 10.0 PRECEDING AND 10.0 FOLLOWING)
FROM (VALUES (1, NAN()), (2, NAN())) t(c0, s0) --- two rows 3, 3
```
In this case, Presto makes the frame to always contain only the peer rows of the current row, even if the frame is logically empty, e.g., range between 10.0 preceding and 20.0 preceding. However, the Velox window operator cannot tell whether NaN in frame bounds comes from the frame offset or the order-by value and will return NULL after this fix.

Would it make sense to disallow NaN in order-by column for window operations as well (given that logically empty frames are not handled correctly)? If not, the window operator will need to know the original frame offsets specified in the SQL queries.

**Comment by aditi-pandit** [comment]
> SELECT SUM(c0) OVER (ORDER BY s0 RANGE BETWEEN 10.0 PRECEDING AND 10.0 FOLLOWING)
> FROM (VALUES (1, NAN()), (2, NAN())) t(c0, s0)

Hi Wei,

  I'm a bit confused by your question. You are saying that generally "range between 10.0 preceding and 20.0 preceding" is an empty frame, but Presto return results like (1, 3).

I see that other systems like DuckDB are also returning similar results

```
duckdb> SELECT SUM(c0) OVER (ORDER BY s0 RANGE BETWEEN 10.0 PRECEDING AND 10.0 FOLLOWING)
   ...> FROM (VALUES (1, CAST('NAN' AS FLOAT)), (2, CAST('NAN' AS FLOAT))) t(c0, s0) ;
┌────────────────────────────────────────────────────────────────────────────┐
│ sum(c0) OVER (ORDER BY s0 RANGE BETWEEN 10.0 PRECEDING AND 10.0 FOLLOWING) │
╞════════════════════════════════════════════════════════════════════════════╡
│                                                                          3 │
│                                                                          3 │
└────────────────────────────────────────────────────────────────────────────┘

duckdb> SELECT SUM(c0) OVER (ORDER BY s0 RANGE BETWEEN 10.0 PRECEDING AND 20.0 PRECEDING)
┌────────────────────────────────────────────────────────────────────────────┐0, s0) ;
│ sum(c0) OVER (ORDER BY s0 RANGE BETWEEN 10.0 PRECEDING AND 20.0 PRECEDING) │
╞════════════════════════════════════════════════════════════════════════════╡
│                                                                          1 │
│                                                                          3 │
└────────────────────────────────────────────────────────────────────────────┘

duckdb> SELECT SUM(c0) OVER (ORDER BY s0 RANGE BETWEEN 20.0 PRECEDING AND 10.0 PRECEDING)
┌────────────────────────────────────────────────────────────────────────────┐0, s0) ;
│ sum(c0) OVER (ORDER BY s0 RANGE BETWEEN 20.0 PRECEDING AND 10.0 PRECEDING) │
╞════════════════════════════════════════════════════════════════════════════╡
│                                                                          1 │
│                                                                          3 │
└────────────────────────────────────────────────────────────────────────────┘
```

So likely they are expected results.

Now coming to Velox:

Your fix is also doing:
```
if constexpr (std::is_floating_point_v<T>) {
      if (data_->isNanAt<T>(partitionRow, mappedFrameRowColumn) &&
          !data_->isNanAt<T>(partitionRow, orderByRowColumn)) {
        validFrames.setValid(currentRow, false);
        continue;
      }
    }
```
So it catches when the frame is NaN but the orderBy is not NaN then mark the frame as empty.. else it uses searchFrameValue NaN in orderBy which should always align to the ends of the partition. Does that not happen right now ? Does it give a value beyond the partition start or end causing the empty frames check to catch those bounds ?

**Comment by kagamiori** [comment]
Hi @aditi-pandit, yeah, let me clarify. I added the condition `&& !data_->isNanAt<T>(partitionRow, orderByRowColumn)` after you approved to handle the situation of order-by column containing NaN specifically. But the condition `if (data_->isNanAt<T>(partitionRow, mappedFrameRowColumn) && !data_->isNanAt<T>(partitionRow, orderByRowColumn))` cannot differentiate whether the NaN in the frame column comes from the order-by column or from the original frame offset in the query. E.g., a query can have NaN both at the order-by column and in frame offsets. In this case, I think the frame is still invalid and we should reject the query. But this is not possible in the current Window operator because the original frame offsets are lost.

Thinking about it twice, I think we can return the same result as Presto for now. And when we reject the queries with NaN frame offsets, we can handle this situation together.

**Comment by aditi-pandit** [comment]
> Hi @aditi-pandit, yeah, let me clarify. I added the condition `&& !data_->isNanAt<T>(partitionRow, orderByRowColumn)` after you approved to handle the situation of order-by column containing NaN specifically. But the condition `if (data_->isNanAt<T>(partitionRow, mappedFrameRowColumn) && !data_->isNanAt<T>(partitionRow, orderByRowColumn))` cannot differentiate whether the NaN in the frame column comes from the order-by column or from the original frame offset in the query. E.g., a query can have NaN both at the order-by column and in frame offsets. In this case, I think the frame is still invalid and we should reject the query. But this is not possible in the current Window operator because the original frame offsets are lost.
> 
> Thinking about it twice, I think we can return the same result as Presto for now. And when we reject the queries with NaN frame offsets, we can handle this situation together.

@kagamiori : In addition to this PR, we also need a Presto side change to reject frames with NaN as a frame bound. This Velox PR is only for cases where NaN appears in the frame column at certain individual rows. But yeah, we cannot distinguish if NaN in frame is coming from ORDER BY or the NaN frame bound.

See previous discussion https://github.com/facebookincubator/velox/pull/11293#issuecomment-2439691172

**Comment by aditi-pandit** [comment]
> Hi @aditi-pandit, yeah, let me clarify. I added the condition `&& !data_->isNanAt<T>(partitionRow, orderByRowColumn)` after you approved to handle the situation of order-by column containing NaN specifically. But the condition `if (data_->isNanAt<T>(partitionRow, mappedFrameRowColumn) && !data_->isNanAt<T>(partitionRow, orderByRowColumn))` cannot differentiate whether the NaN in the frame column comes from the order-by column or from the original frame offset in the query. E.g., a query can have NaN both at the order-by column and in frame offsets. In this case, I think the frame is still invalid and we should reject the query. But this is not possible in the current Window operator because the original frame offsets are lost.
> 
> Thinking about it twice, I think we can return the same result as Presto for now. And when we reject the queries with NaN frame offsets, we can handle this situation together.

@kagamiori : I see. I think the new condition works well.

As far as being able to distinguish if NaN in frame column is from order by or frame column, Velox is not setup that way. We will have to explicitly reject NaN frames in Presto with a separate fix.

The inconsistency leads to issues like https://github.com/facebookincubator/velox/pull/11293#issuecomment-2439691172

**Comment by aditi-pandit** [comment]
@kagamiori : Needn't be in this same PR, but we should add documentation for this NaN behavior in the section on k range frames in the window function developer documentation at https://facebookincubator.github.io/velox/develop/window.html#k-range-frames

**Comment by kagamiori** [comment]
> @kagamiori : Needn't be in this same PR (though better for it), but we should add documentation for this NaN behavior in the section on k range frames in the window function developer documentation at https://facebookincubator.github.io/velox/develop/window.html#k-range-frames

Thanks! I'll update the documentation in a separate PR.

**Pull request closed without merging**
